arangod/RocksDBEngine/RocksDBSha256Checksum.cpp

```diff
@@ -0,0 +1,212 @@
+  EVP_MD_CTX_free(_context);
+#else
+  EVP_MD_CTX_destroy(_context);
+#endif
+}
+
+void RocksDBSha256Checksum::Update(char const* data, size_t n) {
+  EVP_DigestUpdate(_context, static_cast<void const*>(data), n);
+}
+
+void RocksDBSha256Checksum::Finalize() {
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int lengthOfHash = 0;
+  if (EVP_DigestFinal_ex(_context, hash, &lengthOfHash) == 0) {
+    TRI_ASSERT(false);
+  }
+  _checksum = basics::StringUtils::encodeHex(reinterpret_cast<char const*>(&hash[0]), lengthOfHash);
+
+  _shaFileManager->storeShaItems(_fileName, _checksum);
+}
+
+std::string RocksDBSha256Checksum::GetChecksum() const {
+  TRI_ASSERT(!_checksum.empty());
+  return _checksum;
+}
+  
+std::unique_ptr<rocksdb::FileChecksumGenerator> RocksDBSha256ChecksumFactory::CreateFileChecksumGenerator(
+    rocksdb::FileChecksumGenContext const& context) {
+  if (context.requested_checksum_func_name == "RocksDBSha256Checksum") {
+    return std::make_unique<RocksDBSha256Checksum>(context.file_name, _shaFileManager);
+  }
+  return std::make_unique<RocksDBSha256Checksum>(context.file_name, _shaFileManager);
+  // for everything else fall back to CRC32 checksum
+ // LOG_DEVEL << (void*) this <<  " GENERATING CRC";
+ // return std::make_unique<rocksdb::FileChecksumGenCrc32c>(context);
+}
+
+bool RocksDBShaFileManager::storeShaItems(std::string const& fileName, std::string const& checksum) {
+  if (writeShaFile(fileName, checksum)) {
+    MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+    _calculatedHashes.try_emplace(TRI_Basename(fileName), checksum);
+    return true;
+  }
+  return false;
+}
+
+bool RocksDBShaFileManager::writeShaFile(std::string const& fileName, std::string const& checksum) {
+  TRI_ASSERT(TRI_Basename(fileName).size() > 4);
+  TRI_ASSERT(isSstFilename(fileName));
+
+  std::string newFileName = fileName.substr(0, fileName.size() - 4);
+  newFileName += ".sha.";
+  newFileName += checksum;
+  newFileName += ".hash";
+  LOG_TOPIC("80257", DEBUG, arangodb::Logger::ENGINES)
+      << "shaCalcFile: done " << fileName << " result: " << newFileName;
+  auto res = TRI_WriteFile(newFileName.c_str(), "", 0);
+  if (res == TRI_ERROR_NO_ERROR) {
+    return true;
+  }
+
+  LOG_TOPIC("8f7ef", WARN, arangodb::Logger::ENGINES)
+      << "shaCalcFile: TRI_WriteFile failed with " << res << " for " << newFileName;
+  return false;
+}
+
+template <typename T>
+bool RocksDBShaFileManager::isSstFilename(T const& fileName) const {
+  return TRI_Basename(fileName).size() >= 4 &&
+         (fileName.compare(fileName.size() - 4, 4, ".sst") == 0);
+}
+
+void RocksDBShaFileManager::deleteFile(std::string const& pathName) {
+  std::string fileNameBuilder;
+  {
+    MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+    auto it = _calculatedHashes.find(TRI_Basename(pathName));
+    if (it != _calculatedHashes.end()) {
+      fileNameBuilder.append(pathName, 0, pathName.size() - 4);  // append without .sst
```

Comment:
`pathName.size()` must be 4 or greater. Let's add an assertion here or some error handling.

Reply:
Addressed